control/cli.py

```diff
@@ -156,298 +207,1590 @@ def connect(self, host, port, client_key, client_cert, server_cert):
+        argument("--load-balancing-group", "-l", help="Load balancing group", type=int),
+        argument("--size", help="New size in MiB", type=int),
+        argument("--rw-ios-per-second", help="R/W IOs per second limit, 0 means unlimited", type=int),
+        argument("--rw-megabytes-per-second", help="R/W megabytes per second limit, 0 means unlimited", type=int),
+        argument("--r-megabytes-per-second", help="Read megabytes per second limit, 0 means unlimited", type=int),
+        argument("--w-megabytes-per-second", help="Write megabytes per second limit, 0 means unlimited", type=int),
+    ]
+    @cli.cmd(ns_args_list)
+    def ns(self, args):
+        """Namespace commands"""
+        if args.ns_command == "add":
+            return self.ns_add(args)
+        elif args.ns_command == "del":
+            return self.ns_del(args)
+        elif args.ns_command == "resize":
+            return self.ns_resize(args)
+        elif args.ns_command == "list":
+            return self.ns_list(args)
+        elif args.ns_command == "get_io_stats":
+            return self.ns_get_io_stats(args)
+        elif args.ns_command == "change_load_balancing_group":
+            return self.ns_change_load_balancing_group(args)
+        elif args.ns_command == "set_qos":
+            return self.ns_set_qos(args)
+        assert False
+
+    @cli.cmd(ns_args_list)
+    def namespace(self, args):
+        """Namespace commands"""
+        return self.ns(args)
+
+def main_common(client, args):
+    server_address = args.server_address
+    server_port = args.server_port
+    client_key = args.client_key
+    client_cert = args.client_cert
+    server_cert = args.server_cert
+    client.connect(server_address, server_port, client_key, client_cert, server_cert)
+    call_function = getattr(client, args.func.__name__)
+    rc = call_function(args)
+    return rc
+
+def main_test(args):
+    if not args:
+        return None
+    try:
+        i = args.index("--format")
+        del args[i:i + 2]
+    except Exception:
+        pass
+    args = ["--format", "python"] + args
+    client = GatewayClient()
+    parsed_args = client.cli.parser.parse_args(args)
+    if parsed_args.subcommand is None:
+        return None
+
+    return main_common(client, parsed_args)
+
+def main(args=None) -> int:
     client = GatewayClient()
     parsed_args = client.cli.parser.parse_args(args)
     if parsed_args.subcommand is None:
         client.cli.parser.print_help()
-        return 0
-    server_address = parsed_args.server_address
-    server_port = parsed_args.server_port
-    client_key = parsed_args.client_key
-    client_cert = parsed_args.client_cert
-    server_cert = parsed_args.server_cert
-    client.connect(server_address, server_port, client_key, client_cert, server_cert)
-    call_function = getattr(client, parsed_args.func.__name__)
-    call_function(parsed_args)
+        return -1
 
+    return main_common(client, parsed_args)
```

Comment:
I'm a bit concerned by the fact that the cli file has quadrupled its size... I understand that there's a lot of code to unwrap and format the data for human-readable output, but it results in a tightly coupling with the server, and that always complicates cross version support.

As I suggested before, I'd try to reduce this file size by using some other CLI parsing tool, like `click` (I'm almost sure that alone might halve the length of this file) and maybe using some templating engine for textual outputs (`jinja2`, `humanize` for the unit conversion, or `tabulate` as already used here).